src/targets/gpu/gemm_impl.cpp

```diff
@@ -81,198 +95,515 @@ shape transpose_batch(const shape& s, unsigned trans_batch)
+                       [](const shape& x) { return to_gpu(generate_argument(x)); });
+
+        // Get the solutions list in 2 rocBLAS steps:
+        // 1.  Find out how many solutions there are and allocate the array
+        // 2.  Get the solutions
+        //
+        rocblas_int list_size = 0;
+        std::vector<rocblas_int> solution_indices;
+        if(strided_batched)
+        {
+            auto common_args = create_strided_batched_args_common(ctx, input_args);
+            rocblas_invoke(&rocblas_gemm_strided_batched_ex_get_solutions,
+                           common_args,
+                           rocblas_gemm_algo_solution_index,
+                           int8_flag,
+                           nullptr,
+                           &list_size);
+            solution_indices.resize(list_size);
+
+            auto common_sol_args = create_strided_batched_args_common(ctx, input_args);
+            rocblas_invoke(&rocblas_gemm_strided_batched_ex_get_solutions,
+                           common_sol_args,
+                           rocblas_gemm_algo_solution_index,
+                           int8_flag,
+                           solution_indices.data(),
+                           &list_size);
+        }
+        else
+        {
+            auto common_args = create_gemm_ex_args_common(ctx, input_args);
+            rocblas_invoke(&rocblas_gemm_ex_get_solutions,
+                           common_args,
+                           rocblas_gemm_algo_solution_index,
+                           int8_flag,
+                           nullptr,
+                           &list_size);
+            solution_indices.resize(list_size);
+
+            auto common_sol_args = create_gemm_ex_args_common(ctx, input_args);
+            rocblas_invoke(&rocblas_gemm_ex_get_solutions,
+                           common_sol_args,
+                           rocblas_gemm_algo_solution_index,
+                           int8_flag,
+                           solution_indices.data(),
+                           &list_size);
+        }
+
+        double bestTime   = std::numeric_limits<double>::max();
+        double first_time = -1;
+        // Initialize to default solution index
+        rocblas_int bestSol = 0;
+        for(auto sol : solution_indices)
+        {
+            // Define the function to be timed
+            auto run_func = [&]() {
+                run(ctx, input_args, sol);
+                ctx.finish();
+            };
+
+            // Warmup: the first few calls to an op. may not be representative since there is
+            // more time taken initializing caches, etc. so we won't time them.
+            for(int cc = 0; cc < cold_calls; ++cc)
+            {
+                run_func();
+            }
+            double host_time = 0.0;
+
+            for(int hc = 0; hc < hot_calls; ++hc)
+            {
+                ctx.finish();
+                host_time += time<microseconds>(run_func);
```

Comment:
Comments transferred from the old PR:

[pfultz2](https://github.com/pfultz2) [last month](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/pull/1585#discussion_r1127103617)
This should just use the time_op function from the driver. We just need to move it to the gpu lib so it can be reused here.


@[bpickrel](https://github.com/bpickrel) bpickrel [last month](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/pull/1585#discussion_r1128807853)
You'll have to specify what you have in mind for that. gemm_impl doesn't use instances of the Migraphx operation, which is what time_op takes as an argument. Instead, we're making the rocblas calls directly. So we can't reuse it directly--what would we substitute here? (Also, does this change belong in the same PR?)

Reply:
Dont worry about using time_op for now.